RXPatchLib/DirectoryPatcher.cs

```diff
@@ -350,25 +370,109 @@ public DirectoryPatcher(XdeltaPatcher filePatcher, string targetPath, string bac
+
+                        // Lock our tmpActions variable so we have unique access to it now
+                        lock (_tmpActions)
+                        {
+                            if (_tmpActions.Any(k => !k.isComplete && !k.isActive && k.PatchSize > 0))
+                            {
+                                // If an action is not complete, not active and is above zero patch size
+                                thisAction = _tmpActions.DefaultIfEmpty(null).FirstOrDefault(xx => !xx.isComplete && !xx.isActive && xx.PatchSize > 0);
+                            }
+                            else
+                            {
+                                // We have no actions that are above a patch size of zero, do we have any that are zero that still need doing?
+                                if (_tmpActions.Any(l => !l.isComplete && !l.isActive && l.PatchSize == 0))
+                                {
+                                    thisAction = _tmpActions.DefaultIfEmpty(null).FirstOrDefault(xx => !xx.isComplete && !xx.isActive && xx.PatchSize == 0);
+                                }
+                                else
+                                {
+                                    // We're done, break out of our loop to close this thread
+                                    break;
+                                }
+                            }
+
+                            // Grab an action that is not complete, and not active and has a file size
+                            
+                            if (thisAction == null)
+                                continue;
+
+                            thisAction.isActive = true;
+                        }
+
+                        RxLogger.Logger.Instance.Write($"Starting action with file size of {thisAction.PatchSize}");
+                        await thisAction.Execute();
+
+                        // Update progress
+                        if (thisAction.PatchSize == 0)
+                        {
+                            secondPhaseProgress.AdvanceItem(1);
+                            progressCallback(secondPhaseProgress);
+                        }
+                        else
+                        {
+                            progress.AdvanceItem(thisAction.PatchSize);
+                            progressCallback(progress);
+                        }
+                        
+                        // Complete this action
+                        lock (_tmpActions)
+                        {
+                            thisAction.isComplete = true;
+                            thisAction.isActive = false;
+                        }
+                    }
+
+                    RxLogger.Logger.Instance.Write("Background worker terminated");
+                };
+
+                x.RunWorkerAsync();
+            }
+
+            while (_tmpActions.Any(x => !x.isComplete))
+            {
+                await Task.Delay(3000);

```

Comment:
I've shortened the delay, but in terms of making it a list of tasks, i am unsure how to manage the list of tasks in such a way that only allows CPUCount -> MaxxDeltaThreads to run at once, instead of spawning, perhaps for another patch after my C# Tasks knowledge has improved?

Reply:
We can just clean this up later yeah, or I can help you find a solution for it